app/services/proofing/resolution/progressive_proofer.rb

```diff
@@ -86,19 +95,62 @@ def proof_with_threatmetrix_if_needed(
         end
       end
 
-      def proof_resolution(applicant_pii:, timer:)
+      def proof_residential_address_if_needed(
+        applicant_pii:,
+        timer:,
+        double_address_verification:
+      )
+        return residential_address_unnecessary_result unless double_address_verification
+
+        timer.time('residential address') do
+          resolution_proofer.proof(applicant_pii)
+        end
+      end
+
+      def residential_address_unnecessary_result
+        Proofing::AddressResult.new(
+          success: true, errors: {}, exception: nil, vendor_name: 'ResidentialAddressNotRequired',
+        )
+      end
+
+      def resolution_unnecessary_result
+        Proofing::AddressResult.new(
+          success: true, errors: {}, exception: nil, vendor_name: 'ResolutionUnnecessary',
+        )
+      end
+
+      # TK: confirm vendor name with Team Ada
+      def resolution_cannot_pass
+        Proofing::AddressResult.new(
+          success: false, errors: {}, exception: nil, vendor_name: 'ResolutionCannotPass',
+        )
+      end
+
+      def proof_id_address_with_lexis_nexis_if_needed(applicant_pii:, timer:,
+                                                      residential_instant_verify_proof:)
+        return residential_instant_verify_proof if applicant_pii[:same_address_as_id] == 'true'
+        return resolution_cannot_pass unless residential_instant_verify_proof.success?
+
         timer.time('resolution') do
           resolution_proofer.proof(applicant_pii)
         end
       end
 
-      def proof_state_id_if_needed(
+      def proof_id_with_aamva_if_needed(
         applicant_pii:, timer:,
-        resolution_result:,
+        residential_instant_verify_proof:,
+        instant_verify_proof:,
         should_proof_state_id:
       )
-        unless should_proof_state_id && user_can_pass_after_state_id_check?(resolution_result)
-          return out_of_aamva_jurisdiction_result
+        if applicant_pii[:same_address_as_id] == 'false'
```

Comment:
I would appreciate feedback on how to refactor this conditional, because it is gross. Mea culpa!  Ancora imparo!

Reply:
Not sure Ruby, but in java, there are some utility class, that you can do something like BooleanUtil.allTrue(...arrOfBool) or BooleanUtil.anyTrue(..) etc.

More advanced is boolean chaining, you can do it through Predicate
https://docs.oracle.com/javase/8/docs/api/java/util/function/Predicate.html

But difference language need special consideration, Java do boolean shortcut, that's say it stops evaluation of the logical operations as soon as possible, so if one of the condition is depends on some side effect(like function calls), it may not get invoked. Not sure Ruby, but i think it may be similar case.